README.md
## Introduction

`rs-benchmark` is a performance testing tool provided by [RStor Inc](https://rstor.io/) to effectively compare different object storages with different protocols. This tool is an improvement of [s3-benchmark](https://github.com/wasabi-tech/s3-benchmark) developed by Wasabi-Tech.

This tool is provided to independently benchmark the RSTOR Space service provided by RStor Inc, as well as the other major CSP’s object storage services like Amazon S3, Azure Blob Storage, Google Cloud Storage, as well as others that rely on the same protocols.

We decided to open-source `rs-benchmark` because we found out that other s3-benchmark tools don't correctly compute test results: when a `GET` test is performed, the tools only consider as error a bad `HTTP` request, i.e. they don't catch server errors. For example, if the server returns an error, e.g. `404` after requesting a non-existent object, the tools will ignore it and the test is considered passed, thus incorrectly measuring a higher throughput.

## Building

#### Requirements
- `git`
- The Go compiler, `go` version 1.11 or greater
- A reasonably recent Linux distro

#### Install Go compiler on Linux

Ubuntu 18.04

First add the repository:
```
sudo add-apt-repository ppa:gophers/archive && \
sudo apt-get update
```
Then, install the Go package:
```
sudo apt-get install golang-1.11-go
```
Finally, update the `$PATH` variable:
```
echo 'export PATH="/usr/lib/go-1.11/bin:$PATH"' >> ~/.bashrc && \
source ~/.bashrc
```

Debian

On Debian Stretch, you will need to enable backports (https://backports.debian.org/Instructions/) and install with:

```
sudo apt-get -t stretch-backports install golang
```

CentOS 7 / RHEL 7

```
sudo yum install epel-release
sudo yum install go
```

Fedora 29/30

```
sudo dnf install golang
```

If you cannot find a package for your distro, follow the instructions at https://golang.org/doc/install

#### Compiling

Clone the repository with
```
git clone https://github.com/RstorLabs/rs-benchmark.git
```
Now run
```
cd rs-benchmark && go build
```
The executable `rs-benchmark` should now be visible in the project folder.

## Usage

Syntax:
```
./rs-benchmark [OPTIONS]
```
Below are the available command line options to the program:

```
Usage of rs-benchmark:
  -a string
        Access key
  -b string
        Bucket for testing
  -d int
        Duration of each test in seconds (default 60)
  -distribute-keys
        distribute keys over two levels of directories (65536 prefixes) (default true)
  -ip string
        forces all hostnames to resolve to this address (s3v2, s3v4 only)
  -l int
        Number of times to repeat test (default 1)
  -max-retries int
        number of retries on failure (default 0. s3v4 only)
  -multipart
        use multipart
  -multipart-concurrency int
        concurrency to use for multipart requests (default 5)
  -multipart-size string
        Size of the multipart chunks (default "5M")
  -no-cleanup
        do not cleanup inserted data
  -no-get
        do not perform GET part of test
  -pause
        whether to pause between upload and download tests
  -prefix string
        will create objects with key: 'prefix/number' (default "Object")
  -protocol string
        client protocol: s3v2, s3v4, azure, gcp
  -r string
        Region for testing
  -s string
        Secret key
  -t int
        Number of parallel requests to run (default 1)
  -u string
        URL for endpoint with method prefix (e.g. https://s3.rstorcloud.io)
  -v    Verbose error output
  -version
        Show version
  -z string
        Size of objects in bytes with suffix K, M, and G (default "1M")

```

At a bare minimum you need to specify: `-a` (access key), `-s` (secret key), `-b` (test bucket), `-u` (endpoint) and `-protocol` (client protocol).

## Examples

Suppose we want to benchmark the RSTOR Space service, using a bucket named `testbucket`, having object size of 10MB, for 90 seconds, using 4 parallel transfers (the number of files to be uploaded in parallel). Then, we would run

```bash
./rs-benchmark \
    -a ACCESS_KEY \
    -s SECRET_KEY \
    -b testbucket \
    -u https://s3.YOUR_CUSTOMER_NAME.rstorcloud.io \
    -t 4 \
    -z 10M \
    -d 90 \
    -r any \
    -protocol s3v4
```

You can create and manage `ACCESS_KEY` and `SECRET_KEY` by yourself through our Customer UI.


If we wanted to benchmark the object storage service from another provider, the `-u` and `-protocol` flags need to be changed with the appropriate host URL and client protocol. For example, if we wanted to run a similar benchmark on Azure, using the same parameters as before, we would run

```bash
./rs-benchmark \
    -a ACCOUNT_NAME \
    -s ACCOUNT_KEY \
    -b testbucket \
    -u https://ACCOUNT_NAME.blob.core.windows.net \
    -t 4 \
    -z 10M \
    -d 90 \
    -r any \
    -protocol azure
```

The same test against Amazon S3 would be:


```bash
./rs-benchmark \
    -a ACCESS_KEY \
    -s SECRET_KEY \
    -b testbucket \
    -u https://s3.amazonaws.com \
    -t 4 \
    -z 10M \
    -d 90 \
    -r any \
    -protocol s3v4
```

You can test also different combinations of multipart parallel transfers with different chunk sizes. For example, if you wanted to test a 5GB upload with 100MB chunk size and 3 parallel connections (the number of multipart chunks to be uploaded in parallel), against RSTOR Space, you would run:

```bash
./rs-benchmark \
    -a ACCESS_KEY \
    -s SECRET_KEY \
    -b testbucket \
    -u https://s3.YOUR_CUSTOMER_NAME.rstorcloud.io \
    -multipart \
    -multipart-concurrency 3 \
    -multipart-size "100M" \
    -t 4 \
    -z 5G \
    -d 90 \
    -r any \
    -protocol s3v4
```

In this last example, the tool will upload 12 chunks in parallel (`multipart-concurrency = 3` chunks times `threads = 4` files).

To increase accuracy of test results, you can tell `rs-benchmark` to repeat the test multiple times with the option `-l`.

## Additional notes

#### Azure Blob Storage

To get started with Azure benchmarking, first obtain credentials from https://docs.microsoft.com/en-us/azure/storage/blobs/storage-quickstart-blobs-python#copy-your-credentials-from-the-azure-portal. Access key is your account name. The Host URL is in the form of `https://ACCOUNT_NAME.blob.core.windows.net`.

#### Google Cloud Storage

Authentication happens at instance level, so you must run the test from a Google cloud instance, which has been authorized to access the storage. Alternatively, you can use the S3 compatibility layer, with the `s3v4` protocol. 

#### Caveats on multipart

Multipart tests are enabled only for `s3v4` protocol. Azure has a chunk size limit of 128MB, while Google Cloud Platform has a limit of 32 chunks per multipart upload. These limits make an apple-to-apple comparison difficult, therefore the `-multipart-concurrency` parameter is automatically disabled when used in combination with `-protocol azure` and `gcp`.


Changelog.md
### v1.1

- feat: option to distribute writes to multiple prefixes
- feat: option to skip the download or delete steps
- feat: add operations per second
- fix: error when deleting objects in parallel

### v1.0 initial release


main.go
/*
# rs-benchmark - A utility to benchmark object storages
# Copyright (C) 2016-2019 RStor Inc (open-source@rstor.io)
#
# This file is part of rs-benchmark.
#
# rs-benchmark is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# rs-benchmark is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Copyright Header.  If not, see <http://www.gnu.org/licenses/>.
*/

package main

import (
	"bufio"
	"bytes"
	"context"
	"crypto/md5"
	"encoding/binary"
	"encoding/hex"
	"flag"
	"fmt"
	"math/rand"
	"net"
	"net/http"
	"net/url"
	"os"
	"sort"
	"strings"
	"sync"
	"text/tabwriter"
	"time"

	"code.cloudfoundry.org/bytefmt"
	log "github.com/sirupsen/logrus"
)

var (
	objectSize                   uint64
	partSize                     uint64
	objectData                   []byte
	verbose                      bool
	durationSecs, threads, loops int
	successfulUploadIDs          []int
	multipartConcurrency         int
	objPrefix                    string
	distributeKeys               bool
	maxRetries                   int
	noCleanup                    bool
	noGet                        bool
	objects                      uint64
	client                       Uploader
)

const version = "1.1"

func main() {
	var accessKey, secretKey, urlHost, bucket, region, sizeArg, multipartSizeArg string
	var protocol string
	var useMultipart, showVersion bool
	var pauseBetweenPhases bool
	var hostIP string

	// Parse command line
	fl := flag.NewFlagSet("rs-benchmark", flag.ExitOnError)
	fl.StringVar(&accessKey, "a", "", "Access key")
	fl.StringVar(&secretKey, "s", "", "Secret key")
	fl.StringVar(&urlHost, "u", "", "URL for endpoint with method prefix (e.g. https://s3.rstorcloud.io)")
	fl.StringVar(&bucket, "b", "", "Bucket for testing")
	fl.IntVar(&durationSecs, "d", 60, "Duration of each test in seconds")
	fl.Uint64Var(&objects, "objects", 0, "Number of objects to successfully upload/download")
	fl.IntVar(&threads, "t", 1, "Number of parallel requests to run")
	fl.IntVar(&loops, "l", 1, "Number of times to repeat test")
	fl.BoolVar(&verbose, "v", false, "Verbose error output")
	fl.BoolVar(&showVersion, "version", false, "Show version")
	fl.StringVar(&region, "r", "", "Region for testing")
	fl.StringVar(&protocol, "protocol", "", "client protocol: s3v2, s3v4, azure, gcp")
	fl.BoolVar(&useMultipart, "multipart", false, "use multipart")
	fl.IntVar(&multipartConcurrency, "multipart-concurrency", 5, "concurrency to use for multipart requests")
	fl.BoolVar(&pauseBetweenPhases, "pause", false, "whether to pause between upload and download tests")
	fl.BoolVar(&noCleanup, "no-cleanup", false, "do not cleanup inserted data")
	fl.BoolVar(&noGet, "no-get", false, "do not perform GET part of test")
	fl.StringVar(&hostIP, "ip", "", "forces all hostnames to resolve to this address (s3v2, s3v4 only)")
	fl.StringVar(&objPrefix, "prefix", "Object", "will create objects with key: 'prefix/number'")
	fl.BoolVar(&distributeKeys, "distribute-keys", true, "distribute keys over two levels of directories (65536 prefixes)")
	fl.IntVar(&maxRetries, "max-retries", 0, "number of retries on failure (default 0. s3v4 only)")
	fl.StringVar(&sizeArg, "z", "1M", "Size of objects in bytes with suffix K, M, and G")
	fl.StringVar(&multipartSizeArg, "multipart-size", "5M", "Size of the multipart chunks")

	switch err := fl.Parse(os.Args[1:]); err {
	case flag.ErrHelp:
		fmt.Println("Available arguments:")
		fl.PrintDefaults()
		fmt.Println("")
		os.Exit(0)
	case nil:
	default:
		fmt.Printf("Unable to parse flags %v\n", err)
	}

	// Check the arguments
	if showVersion == true {
		fmt.Printf("RSTOR rs-benchmark v%s.\n\n", version)
		os.Exit(0)
	}

	fmt.Printf("rs-benchmark v%s - a compact tool for benchmarking different object storages\n", version)
	fmt.Println("Copyright (C) 2016-2020 RStor Inc (open-source@rstor.io)")
	fmt.Println("Released under GPL v3 license")
	fmt.Println()

	if protocol == "" {
		fmt.Println("Missing argument -protocol for client protocol.")
		printHelpAndExit()
	}

	if protocol == "s3v4" && region == "" {
		fmt.Println("Protocol s3v4 requires the region to be specified.")
		printHelpAndExit()
		os.Exit(-1)
	}

	hostIPForPrinting := ""
	if hostIP == "" && urlHost == "" {
		fmt.Println("Missing host information.")
		printHelpAndExit()
	}

	if hostIP != "" {
		dTransport := httpClient.Transport.(*http.Transport)

		dTransport.DialContext = func(ctx context.Context, network, addr string) (net.Conn, error) {
			addr = hostIP
			return dialer.DialContext(ctx, network, addr)
		}

		hostIPForPrinting = hostIP
	} else {
		u, err := url.Parse(urlHost)
		if err != nil {
			fmt.Println("Invalid url ", err)
			printHelpAndExit()
		}

		host := strings.Split(u.Host, ":")[0]
		ips, err := net.LookupIP(host)
		if err != nil {
			fmt.Println("Can't resolve host ", u.Host)
			printHelpAndExit()
		} else {
			var ipStrings []string
			for _, ip := range ips {
				ipStrings = append(ipStrings, ip.String())
			}
			hostIPForPrinting = strings.Join(ipStrings, ", ")
		}
	}

	if protocol != "gcp" {
		if accessKey == "" {
			fmt.Println("Missing argument -a for access key.")
			printHelpAndExit()
		}
		if secretKey == "" {
			fmt.Println("Missing argument -s for secret key.")
			printHelpAndExit()
		}
	}

	if bucket == "" {
		fmt.Println("Missing argument -b for bucket.")
		printHelpAndExit()
	}

	var err error

	if objectSize, err = bytefmt.ToBytes(sizeArg); err != nil {
		fmt.Printf("Invalid -z argument for object size: %v\n", err)
		printHelpAndExit()
	}

	if partSize, err = bytefmt.ToBytes(multipartSizeArg); err != nil {
		fmt.Printf("Invalid -multipart-size argument for part size: %v\n", err)
		printHelpAndExit()
	}

	if multipartConcurrency <= 0 {
		multipartConcurrency = 1
	}

	if maxRetries < 0 {
		maxRetries = 0
	}

	switch protocol {
	case "s3v4":
		v4Client := NewS3AwsV4(accessKey, secretKey, urlHost, region)
		v4Client.UseMultipart = useMultipart
		client = v4Client
	case "s3v2":
		if useMultipart {
			fmt.Println("Multipart not supported")
			printHelpAndExit()
		}
		if region != "" {
			fmt.Println("-region not supported for s3v2. Drop option.")
			printHelpAndExit()
		}
		client = NewS3AwsV2(accessKey, secretKey, urlHost)
	case "azure":
		if region != "" {
			fmt.Println("region param not supported")
			printHelpAndExit()
		}
		if useMultipart && multipartConcurrency > 1 {
			fmt.Println("Multipart concurrency is fixed to one")
			multipartConcurrency = 1
		}
		aup := NewAzureUploader(accessKey, secretKey, urlHost)
		aup.UseMultipart = useMultipart
		client = aup
	case "gcp":
		if region != "" {
			fmt.Println("region param not supported")
			printHelpAndExit()
		}
		if useMultipart && multipartConcurrency > 1 {
			fmt.Println("Multipart concurrency is fixed to one")
			multipartConcurrency = 1
		}
		gup := NewGCP()
		gup.UseMultipart = useMultipart
		client = gup
	default:
		fmt.Println("unknown client type. Available: s3v4, s3v2, azure, gpc")
		printHelpAndExit()
	}
	fmt.Println("Benchmark parameters:")

	tw := tabwriter.NewWriter(os.Stdout, 16, 2, 1, ' ', 0)
	tabLine := func(name string, value interface{}) {
		_, _ = fmt.Fprint(tw, name, "\t", value, "\n")
	}
	tabLine("Endpoint URL", urlHost)
	tabLine("Protocol", protocol)
	tabLine("Host ip", hostIPForPrinting)
	if region != "" {
		tabLine("Region", region)
	}
	tabLine("Bucket", bucket)
	tabLine("Prefix", objPrefix)
	tabLine("Distribute keys", distributeKeys)
	tabLine("Test time", durationSecs)
	tabLine("Test objects", objects)
	tabLine("Threads", threads)
	tabLine("Size", sizeArg)
	tabLine("Loops", loops)
	if useMultipart {
		tabLine("Multipart",
			fmt.Sprintf("%s per part, %d parallel uploads", multipartSizeArg, multipartConcurrency))
	} else {
		tabLine("Multipart", "false")
	}
	tabLine("Max retries", maxRetries)
	_ = tw.Flush()
	fmt.Println()

	// Test access to the bucket
	err = client.Prepare(bucket)
	if err != nil {
		log.Fatalf("error preparing the bucket: %v\n" +
			"For more information, run again with flag -v.")
	}

	// Initialize data for the bucket
	objectData = make([]byte, objectSize)
	rand.Read(objectData)

	// Loop running the tests
	for loop := 1; loop <= loops; loop++ {
		runLoop(loop, pauseBetweenPhases)
	}

	fmt.Println("\nDone.")
}

func runLoop(loop int, pauseBetweenPhases bool) {
	if loop > 1 && pauseBetweenPhases {
		fmt.Printf("Loop %d done\n", loop-1)
		pause()
	}

	fmt.Printf("Starting loop %d...\n", loop)

	// Run the upload case
	successfulUploadIDs = make([]int, 0, 1000) // reused in download step
	totalDuration := float64(0)
	indexes := make(chan int, threads)
	res := make(chan TransferResult, threads)

	ctx, cancelRemainingUploads := context.WithCancel(context.Background())
	for n := 0; n <= threads; n++ {
		go runUpload(ctx, indexes, res)
	}

	startTime := time.Now()
	uploads := runAndCollectResults(indexes, res)
	cancelRemainingUploads()
	uploadTime := time.Now().Sub(startTime).Seconds()

	var uploadedBytes uint64
	var uploadDurations []float64
	for _, v := range uploads {
		if v.Error != nil {
			continue
		}
		successfulUploadIDs = append(successfulUploadIDs, v.Id)
		uploadedBytes += objectSize
		uploadDurations = append(uploadDurations, v.Duration.Seconds())
		totalDuration += v.Duration.Seconds()
	}
	sort.Float64s(uploadDurations)

	// log.Info(successfulUploadIDs)
	successfulUploads := len(successfulUploadIDs)
	failedUploads := len(uploads) - len(successfulUploadIDs)
	uploadMBps := (float64(uploadedBytes) / uploadTime) / (1000 * 1000)
	uploadPerSecond := float64(successfulUploads) / uploadTime

	fmt.Printf("%-9s%-6s%-11s%-7s%-12s%-8s%-8s%-12s\n",
		"Threads", "Size", "Operation", "Time", "Successful", "Failed", "MBps", "OPps")
	fmt.Printf("%-9d%-6v%-11s%-7.2f%-12v%-8v%-8.2f%-12.2f\n",
		threads, bytefmt.ByteSize(objectSize), "PUT", uploadTime, successfulUploads, failedUploads, uploadMBps, uploadPerSecond)

	if len(successfulUploadIDs) < 5 {
		if verbose == false {
			log.Fatal("Not enough successful uploads to continue. For more information, run again with flag -v")
		} else {
			log.Fatal("Not enough successful uploads to continue.")
		}
	}

	if pauseBetweenPhases {
		pause()
	}

	// Run the download case
	indexes = make(chan int, threads)
	res = make(chan TransferResult, 10)

	ctx, cancelRemainingDownloads := context.WithCancel(context.Background())
	for n := 0; n <= threads; n++ {
		go runDownload(ctx, indexes, res)
	}

	startTime = time.Now()
	downloads := runAndCollectResults(indexes, res)
	cancelRemainingDownloads()
	downloadTime := time.Now().Sub(startTime).Seconds()

	var successfulDownloads int
	var failedDownloads int
	var downloadedBytes uint64
	totalDuration = 0
	var downloadDurations []float64
	for _, d := range downloads {
		if d.Error != nil {
			failedDownloads++
		} else {
			successfulDownloads++
			downloadedBytes += objectSize
			totalDuration += d.Duration.Seconds()
			downloadDurations = append(downloadDurations, d.Duration.Seconds())
		}
	}
	sort.Float64s(downloadDurations)

	if successfulDownloads == 0 {
		log.Fatal("All downloads failed\n" +
			"For more information, run again with flag -v.")
	}

	mbPs := (float64(downloadedBytes) / downloadTime) / (1000 * 1000)
	downloadsPerSecond := float64(successfulDownloads) / downloadTime

	fmt.Printf("%-9d%-6v%-11s%-7.2f%-12v%-8v%-8.2f%-12.2f\n",
		threads, bytefmt.ByteSize(objectSize), "GET", downloadTime, successfulDownloads, failedDownloads, mbPs, downloadsPerSecond)

	if noCleanup {
		fmt.Println("Not performing cleanup, as requested")
		return
	}

	if pauseBetweenPhases {
		pause()
	}

	ctx = context.Background()
	fmt.Println("Deleting test objects")
	deleteChan := make(chan int)
	deleteWg := sync.WaitGroup{}
	deleteWg.Add(threads)
	for n := 0; n < threads; n++ {
		go func() {
			defer deleteWg.Done()

			for idx := range deleteChan {
				_ = client.DoDelete(ctx, objectKey(idx))
			}
		}()
	}
	for i, v := range successfulUploadIDs {
		deleteChan <- v
		if i > 0 && i%10000 == 0 {
			fmt.Printf("%d deletes completed\n", i)
		}
	}
	close(deleteChan)
	deleteWg.Wait()
}

func runAndCollectResults(indexes chan<- int, res <-chan TransferResult) []TransferResult {
	var (
		nextId         int
		successUploads uint64
	)

	for nextId = 0; nextId < threads+1; nextId++ {
		indexes <- nextId
	}

	results := make([]TransferResult, 0, objects)
	deadline := time.After(time.Second * time.Duration(durationSecs))

Loop:
	for {
		select {
		case <-deadline:
			break Loop
		case r := <-res:
			results = append(results, r)
			if r.Error != nil {
				indexes <- r.Id
			} else {
				successUploads++
				// break out after N objects
				if successUploads == objects {
					break Loop
				}
				indexes <- nextId
				nextId++
			}
		}
	}

	return results
}

type TransferResult struct {
	Id       int
	Duration time.Duration
	Error    error
}

func runUpload(ctx context.Context, ids <-chan int, res chan<- TransferResult) {
	for id := range ids {
		reader := bytes.NewReader(objectData)

		startTime := time.Now()
		r := client.DoUpload(ctx, objectKey(id), reader)

		r.Duration = time.Now().Sub(startTime)
		r.Id = id

		if r.Error != nil {
			if !strings.Contains(r.Error.Error(), "context canceled") {
				log.Error(r.Error)
			}
		}
		res <- r
	}
}

func objectKey(idx int) string {
	if !distributeKeys {
		return fmt.Sprintf("%s/%d", objPrefix, idx)
	}

	// hash the index
	bin := make([]byte, 8)
	binary.LittleEndian.PutUint64(bin, uint64(idx))
	sum := md5.Sum(bin)
	hexStr := hex.EncodeToString(sum[:])

	// format hex string to xx/yy/zzzzzzzzzzzzzzzzzzzzzzzzzzzz
	b := strings.Builder{}
	b.WriteString(objPrefix)
	b.WriteString("/")
	b.WriteString(hexStr[:2])
	b.WriteString("/")
	b.WriteString(hexStr[2:4])
	b.WriteString("/")
	b.WriteString(hexStr[4:])

	return b.String()
}

func runDownload(ctx context.Context, indexes <-chan int, res chan<- TransferResult) {
	for id := range indexes {
		idx := successfulUploadIDs[id%len(successfulUploadIDs)]

		startTime := time.Now()
		r := client.DoDownload(ctx, objectKey(idx))

		r.Duration = time.Now().Sub(startTime)
		r.Id = id

		if r.Error != nil {
			if !strings.Contains(r.Error.Error(), "context canceled") {
				log.Error(r.Error)
			}
		}
		res <- r
	}
}

func pause() {
	fmt.Print("Press 'Enter' to continue to the next phase")
	_, _ = bufio.NewReader(os.Stdin).ReadBytes('\n')
}

func printHelpAndExit() {
	fmt.Print("Abort.\n\nRun \"./rs-benchmark --help\" for usage.\n")
	os.Exit(-1)
}
